pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2014-2018 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.helix.core.periodictask;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.core.periodictask.BasePeriodicTask;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The base periodic task for pinot controller only. It uses <code>PinotHelixResourceManager</code> to determine
+ * which table resources should be managed by this Pinot controller.
+ */
+public abstract class ControllerPeriodicTask extends BasePeriodicTask {
+  public static final Logger LOGGER = LoggerFactory.getLogger(ControllerPeriodicTask.class);
+  protected final PinotHelixResourceManager _pinotHelixResourceManager;
+  private boolean _amILeader;
+  private static final int DEFAULT_INITIAL_DELAY_IN_SECOND = 120;
+
+  public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds,
+      PinotHelixResourceManager pinotHelixResourceManager) {
+    this(taskName, runFrequencyInSeconds, DEFAULT_INITIAL_DELAY_IN_SECOND, pinotHelixResourceManager);
+  }
+
+  public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds, long initialDelayInSeconds,
+      PinotHelixResourceManager pinotHelixResourceManager) {
+    super(taskName, runFrequencyInSeconds, initialDelayInSeconds);
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    setAmILeader(false);
+  }
+
+  @Override
+  public void init() {
+  }
+
+  @Override
+  public void run() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      skipLeaderTask();
+    } else {
+      List<String> allTableNames = _pinotHelixResourceManager.getAllTables();
```

Comment:
I would suggest adding the iterator to the base class, and handling any exception here after calling a method in the sub-class to process a single table. We have had bugs before where exceptions were not handled, so the periodic tasks stopped at the table that encountered the exception.  Ideally the state should be maintained in the base class and the sub-class methods invoked when the state changes. It also gets easy to unit test sub-classes with this framework.

Reply:
I think what you're concerned about is the logic in `retention manager`, where the retention manager stopped if any exception occurred in **one** table. We can handle it in another PR.

Reply:
I am concerned about any new task we add failing to handle exceptions, and in genral, about unifying the task handling as much as possible. I am fine doing this in another PR.